### PR TITLE
[clang][analyzer] Support 'tello' and 'fseeko' in the StreamChecker

### PR DESCRIPTION
**Pull request by benshi001**

### CODE REVIEW

**Comment by balazske** [comment]
It is better to replace in code of `evalFtell` the `C.getASTContext().LongTy` with `Call.getResultType()` (occurs 2 times), this is more exact if `off_t` is defined to something else than `long`.

File: clang/test/Analysis/stream-error.c [hunk 1]
```diff
@@ -324,6 +355,57 @@ void error_fseek_0(void) {
...
+  int rc = fseeko(F, 0, SEEK_SET);
+  if (rc) {
+    int IsFEof = feof(F), IsFError = ferror(F);
+    // Get ferror or no error, but not feof.
+    clang_analyzer_eval(IsFError);
+    // expected-warning@-1 {{FALSE}}
+    // expected-warning@-2 {{TRUE}}
+    clang_analyzer_eval(IsFEof);
+    // expected-warning@-1 {{FALSE}}
+    // Error flags should not change.
```
Comment (balazske) [inline clang/test/Analysis/stream-error.c:371]: Part after "error flags should not change" is not needed here, it is enough to have this at one place.

File: clang/test/Analysis/stream-error.c [hunk 2]
```diff
@@ -324,6 +355,57 @@ void error_fseek_0(void) {
...
+    clang_analyzer_eval(feof(F)); // expected-warning {{FALSE}}
+    if (IsFError)
+      clang_analyzer_eval(ferror(F)); // expected-warning {{TRUE}}
+    else
+      clang_analyzer_eval(ferror(F)); // expected-warning {{FALSE}}
```
Comment (balazske) [inline clang/test/Analysis/stream-error.c:376]: Part after "error flags should not change" is not needed here, it is enough to have this at one place.
```suggestion
```

File: clang/test/Analysis/stream-error.c [hunk 3]
```diff
@@ -324,6 +355,57 @@ void error_fseek_0(void) {
...
+  } else {
+    clang_analyzer_eval(feof(F));   // expected-warning {{FALSE}}
+    clang_analyzer_eval(ferror(F)); // expected-warning {{FALSE}}
+    // Error flags should not change.
+    clang_analyzer_eval(feof(F));   // expected-warning {{FALSE}}
+    clang_analyzer_eval(ferror(F)); // expected-warning {{FALSE}}
```
Comment (balazske) [inline clang/test/Analysis/stream-error.c:382]:
```suggestion
```

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 4]
```diff
@@ -268,8 +268,12 @@ class StreamChecker : public Checker<check::PreCall, eval::Call,
         std::bind(&StreamChecker::evalUngetc, _1, _2, _3, _4), 1}},
       {{{"fseek"}, 3},
```
Comment (balazske) [inline clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:269]:
```suggestion
      {{{"fseek", "fseeko"}, 3},
```
The same is applicable to `fopen` and `fdopen` in `CallDescriptionMap`.

**Comment by benshi001** [comment]
> `Call.getResultType()`

This is fixed. Thanks!

**Comment by benshi001** [comment]
> It is better to replace in code of `evalFtell` the `C.getASTContext().LongTy` with `Call.getResultType()` (occurs 2 times), this is more exact if `off_t` is defined to something else than `long`.

This has been fixed. Thanks!

File: clang/test/Analysis/stream-error.c [hunk 5]
```diff
@@ -324,6 +343,60 @@ void error_fseek_0(void) {
...
+void error_ftello(void) {
+  FILE *F = tmpfile();
+  if (!F)
+    return;
+  long rc = ftello(F);
+  if (rc >= 0)
+    clang_analyzer_warnIfReached();          // expected-warning {{REACHABLE}}
+  else
+    clang_analyzer_eval(rc == -1);           // expected-warning {{TRUE}}
+  clang_analyzer_eval(feof(F) && ferror(F)); // expected-warning {{FALSE}}
```
Comment (balazske) [inline clang/test/Analysis/stream-error.c:396]: To be exact, this test should be a copy of the previous test, only with `ftello` (and use of `off_t`).

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp (hunk 4 above)
Comment (balazske) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:269]: This is not fixed yet, `fseek` and `fseeko` (and `ftell`, `ftello`) should be movable to one entry.

**Review by balazske — APPROVED** [review]

**Pull request merged**